In [1]:
from bs4 import BeautifulSoup
import urllib.request # url lib은, 위에서 배운 request와 비슷한데, url을 쉽게 다룰 수 있도록 함.
import re

url = 'https://www.hollys.co.kr/store/korea/korStore2.do'

# urllib.request: urllib 모듈의 하위 모듈로, 웹 페이지에 HTTP 요청을 보내고 응답을 받을 수 있는 기능을 제공
# urlopen 지정된 URL에 HTTP 요청을 보내고 서버로부터 응답을 받아오는 함수, 변수에는 requests와 비슷한 형식의 객체가 담김.
html = urllib.request.urlopen(url)

soup = BeautifulSoup(html, 'html.parser')

# css path에 대해서 설명
tags = soup.select('html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content fieldset fieldset div.tableType01 table.tb_store tbody tr td.center_t a')

store_numbers = []

for tag in tags:
    onclick_text = tag.get('onclick', '')

    # r'storeView\((\d+)\)': 이 패턴은 storeView(숫자) 형식을 찾기 위해 사용됨.
    match = re.search(r'storeView\((\d+)\)', onclick_text)
    if match:
        store_number = match.group(1)
        store_numbers.append(store_number)

print(store_numbers)

['1151', '1151', '1150', '1150', '1148', '1148', '1146', '1146', '1145', '1145', '1144', '1144', '1142', '1142', '1141', '1141', '1138', '1138', '1137', '1137']


In [4]:
# 중복되는 값이 있으니, set으로 중복값 제거.

unique_store_numbers = list(set(store_numbers))
print(unique_store_numbers)

['1142', '1150', '1137', '1151', '1144', '1146', '1138', '1148', '1141', '1145']


https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1151

위에서 가져온 1151이라는 값은, 페이지 1에 브랜치id 1151이라는 링크를 의미.

즉, 페이지 1에는 len(unique_store_numbers)만큼의 매장 아이디가 있고, 해당 길이를 넘어가면, page2로 이동해서 위의 내용을 반복해야 함.


In [6]:
page_number = 1

for branch_number in unique_store_numbers:
  store_url = f"https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={page_number}&branchId={branch_number}"
  print(store_url)


https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1142
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1150
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1137
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1151
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1144
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1146
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1138
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1148
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1141
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1145


몇페이지까지 있는지 직접 확인

할리스는 n 페이지까지 있기에 1부터 50까지 반복하면 됨.


In [ ]:
page_number = 1

for branch_number in unique_store_numbers:
  store_url = f"https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={page_number}&branchId={branch_number}"
  print(store_url)


In [29]:
from bs4 import BeautifulSoup
import urllib.request 
import re

for page_number in range(1,5):
  url = f"https://www.hollys.co.kr/store/korea/korStore2.do?pageNo={page_number}&sido=&gugun=&store="
  
  print(url)

  html = urllib.request.urlopen(url)

  soup = BeautifulSoup(html, 'html.parser')

  tags = soup.select('html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content fieldset fieldset div.tableType01 table.tb_store tbody tr td.center_t a')

  store_numbers = []

  for tag in tags:
      onclick_text = tag.get('onclick', '')

      # r'storeView\((\d+)\)': 이 패턴은 storeView(숫자) 형식을 찾기 위해 사용됨.
      match = re.search(r'storeView\((\d+)\)', onclick_text)
      if match:
          store_number = match.group(1)
          store_numbers.append(store_number)

  print(store_numbers)
  for branch_number in unique_store_numbers:
    store_url = f"https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={page_number}&branchId={branch_number}"
    print(store_url)

https://www.hollys.co.kr/store/korea/korStore2.do?pageNo=1&sido=&gugun=&store=
['1151', '1151', '1150', '1150', '1148', '1148', '1146', '1146', '1145', '1145', '1144', '1144', '1142', '1142', '1141', '1141', '1138', '1138', '1137', '1137']
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1142
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1150
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1137
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1151
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1144
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1146
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1138
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1148
https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1141
https://www.hollys.co.kr/store/korea/korStoreDeta

In [30]:
tags = soup.select('html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content div.store_view01 div.store_detail02 ul li span.center_t')

# css 경로가 같은 경우가 있기에, css 경로가 같은 모든 경우를 가져와서 list에 넣도록 함.
texts = [tag.get_text(strip=True) for tag in tags]

# 결과 출력
print(texts)

[]


In [33]:
import requests
from bs4 import BeautifulSoup

store_url = "https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo=1&branchId=1144"
response = requests.get(store_url)

# 실전에선 이렇게 예외처리 하는 것이 상당히 중요함.
if response.status_code != 200:
    print(f"Failed to retrieve the webpage. Status code: {response.status_code}")

soup = BeautifulSoup(response.text, 'html.parser')

# CSS 선택자 문자열입니다.
store_info_css = "html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content div.store_view01 div.store_detail02 ul li span.center_t"

# CSS 선택자를 사용하여 태그를 선택합니다.
store_info_tags = soup.select(store_info_css)

# 선택된 태그가 존재하는지 확인.
if not store_info_tags:
    print("cant find store_info_tags")
else:
    texts = [tag.get_text(strip=True) for tag in store_info_tags]

    # 결과 출력
    print(texts)
    print(texts[3])



['경상북도 김천시 혁신로 246 (율곡동) 1층', '.', '054-433-4333', '09:00~22:00', '', '무료주차']
09:00~22:00


In [35]:
from bs4 import BeautifulSoup
import urllib.request 
import re

for page_number in range(1,5):
  url = f"https://www.hollys.co.kr/store/korea/korStore2.do?pageNo={page_number}&sido=&gugun=&store="
  
  
  html = urllib.request.urlopen(url)

  soup = BeautifulSoup(html, 'html.parser')

  tags = soup.select('html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content fieldset fieldset div.tableType01 table.tb_store tbody tr td.center_t a')

  store_numbers = []

  for tag in tags:
      onclick_text = tag.get('onclick', '')

      match = re.search(r'storeView\((\d+)\)', onclick_text)
      if match:
          store_number = match.group(1)
          store_numbers.append(store_number)

  print(store_numbers)
  for branch_number in unique_store_numbers:
    store_url = f"https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={page_number}&branchId={branch_number}"



    if response.status_code != 200:
        print(f"Failed to retrieve the webpage. Status code: {response.status_code}")

    soup = BeautifulSoup(response.text, 'html.parser')

    # CSS 선택자 문자열입니다.
    store_info_css = "html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content div.store_view01 div.store_detail02 ul li span.center_t"

    # CSS 선택자를 사용하여 태그를 선택합니다.
    store_info_tags = soup.select(store_info_css)

    # 선택된 태그가 존재하는지 확인.
    if not store_info_tags:
        print("cant find store_info_tags")
    else:
        texts = [tag.get_text(strip=True) for tag in store_info_tags]

        print(texts[3])

['1151', '1151', '1150', '1150', '1148', '1148', '1146', '1146', '1145', '1145', '1144', '1144', '1142', '1142', '1141', '1141', '1138', '1138', '1137', '1137']
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
['1136', '1136', '1135', '1135', '1134', '1134', '1133', '1133', '1131', '1131', '1130', '1130', '1129', '1129', '1128', '1128', '1126', '1126', '1125', '1125']
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
['1124', '1124', '1123', '1123', '1122', '1122', '1121', '1121', '1120', '1120', '1119', '1119', '1118', '1118', '1117', '1117', '1116', '1116', '1115', '1115']
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
09:00~22:00
['1114', '1114', '1113', '1113', '1112', '1112', '1111', '1111', '1110', '1110', '1109', '1109', '1108', '1108', '1107', '1107', '1106', '1106', '1105', '110

매장 명과 함께 출력


In [52]:
from bs4 import BeautifulSoup
import re

for page_number in range(1,5):
    url = f"https://www.hollys.co.kr/store/korea/korStore2.do?pageNo={page_number}&sido=&gugun=&store="

    response = requests.get(store_url)

    print(url)
    #   html = urllib.request.urlopen(url)

    soup = BeautifulSoup(response.text, 'html.parser')

    tags = soup.select('html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content fieldset fieldset div.tableType01 table.tb_store tbody tr td.center_t a')

    store_numbers = []

    for tag in tags:
        onclick_text = tag.get('onclick', '')

        match = re.search(r'storeView\((\d+)\)', onclick_text)
        if match:
            store_number = match.group(1)
            store_numbers.append(store_number)

    print("store_numbers", store_numbers)

    for branch_number in unique_store_numbers:
        store_url = f"https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={page_number}&branchId={branch_number}"

        # print(store_url)

        response = requests.get(store_url)


        if response.status_code != 200:
            print(f"Failed to retrieve the webpage. Status code: {response.status_code}")
            continue # 다음 매장으로 넘어감

        soup = BeautifulSoup(response.text, 'html.parser')

        store_info_css = "html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content div.store_view01 div.store_detail02 ul li span.center_t"

        store_name_css = "html body div.wrap div#contentsWr.contents_wr div#contents.contents.po_design div.content div.store_view01 div.store_detail02 p.store_name"

        store_info_tags = soup.select(store_info_css)
        store_name_tags = soup.select(store_name_css)

        if not store_info_tags:
            print("cant find store_info_tags")
        else:
            info = [tag.get_text(strip=True) for tag in store_info_tags]

            name = [tag.get_text(strip=True) for tag in store_name_tags]

            # print(f"{store_name_tags}, {info[3]}")
            print(f"{store_name_tags[0].text}, {info[3]}")

https://www.hollys.co.kr/store/korea/korStore2.do?pageNo=1&sido=&gugun=&store=
store_numbers []
평촌점, 월~토 06~03,일 06~24
인천공항교통센터점, 06:00 ~22:00
왕십리역점, 월-일 10:00 ~ 22:00
대구팔달교DT점, 09:00 ~ 23:00
김천혁신도시점, 09:00~22:00
대구장기점, 09:00 ~ 24:00
구미CGV점, 09:00~21:00
제주애월천덕로점, 09:00~19:00
남동케이원점 , 08:00~20:00 (일요일휴무) (공휴일 10:00-17:00)
부산연산점, 08:30 ~ 22:00
https://www.hollys.co.kr/store/korea/korStore2.do?pageNo=2&sido=&gugun=&store=
store_numbers []
평촌점, 월~토 06~03,일 06~24
인천공항교통센터점, 06:00 ~22:00
왕십리역점, 월-일 10:00 ~ 22:00
대구팔달교DT점, 09:00 ~ 23:00
김천혁신도시점, 09:00~22:00
대구장기점, 09:00 ~ 24:00
구미CGV점, 09:00~21:00
제주애월천덕로점, 09:00~19:00
남동케이원점 , 08:00~20:00 (일요일휴무) (공휴일 10:00-17:00)
부산연산점, 08:30 ~ 22:00
https://www.hollys.co.kr/store/korea/korStore2.do?pageNo=3&sido=&gugun=&store=
store_numbers []
평촌점, 월~토 06~03,일 06~24
인천공항교통센터점, 06:00 ~22:00
왕십리역점, 월-일 10:00 ~ 22:00
대구팔달교DT점, 09:00 ~ 23:00
김천혁신도시점, 09:00~22:00
대구장기점, 09:00 ~ 24:00
구미CGV점, 09:00~21:00
제주애월천덕로점, 09:00~19:00
남동케이원점 , 08:00~20:00 (일요일휴무) (공휴일 10

위에 코드는 조금 더러움.

코드 리팩토링을 진행하면, 즉, 코드를 더욱 깔끔하게 한다면..


1. url을 반복문 바깥에서 선언하고, 중괄호를 통해서 변수를 넣을 공간만 만들어주는방식으로 진행할 수 있음.

```python
base_url = "https://www.hollys.co.kr/store/korea/korStore2.do?pageNo={}&sido=&gugun=&store="
detail_base_url = "https://www.hollys.co.kr/store/korea/korStoreDetail2.do?pageNo={}&branchId={}"
```

이렇게..

<br>

2. css path 전체를 가져오지 않고, 마지막 태그만 가져오는 방식을 사용해도 되는 경우가 꽤 있음.

```python
tags = soup.select('td.center_t a')
```

이름이 중복되지만 않으면 되기에, 이렇게 해도 아마 잘 작동할 거임.

<br>

3. 예외처리를 진행하는 것이 나중에 디버깅을 할 때 상당히 도움이 됨.

```python
if response.status_code != 200:
    print(f"error, status code: {response.status_code}")
    continue
```

반복문 안에 이걸 넣는다면, 해당 response가 정상적이지 않을 때, 다음 response로 넘어가게 됨.
